nbgrader_to_canvas/upload_grades.py

```diff
@@ -0,0 +1,379 @@
+                        nb_student_and_score['posted_grade'] = nb_submission.score*(100-int(self._late_penalty))/100
+                    else:
+                        nb_student_and_score['posted_grade'] = nb_submission.score
+
+                    if nb_student.id in self.nbgrader_feedback:
+                        nb_student_and_score['file_ids'] = [self.nbgrader_feedback[nb_student.id]]
+                        nb_student_and_score['text_comment'] = 'Transferred from datahub.'
+
+                except MissingEntry:
+                    nb_student_and_score['posted_grade'] = ''
+
+
+                # student.id will give us student's username, ie shrakibullah. we will need to compare this to
+                # canvas's login_id instead of user_id              
+
+                # convert nbgrader username to canvas id (integer)
+                canvas_student_id = self.canvas_students[nb_student.id]
+                nb_grade_data[canvas_student_id] = nb_student_and_score
+                counter += 1
+                update_status(self._form_nb_assign_name, self._course_id, completion=35+30*counter/self._num_students)
+            return nb_grade_data
+
+    def _get_assignment(self):
+        assignment = self._course.get_assignment(self._form_canvas_assign_id)
+        if assignment is None:
+            raise Exception('Invalid form canvas_assign_id')
+        return assignment
+
+    def _get_max_score(self):
+        with Gradebook("sqlite:////mnt/nbgrader/"+self._nbgrader_course+"/grader/gradebook.db") as gb:
+            nb_assignment = gb.find_assignment(self._form_nb_assign_name)
+            return nb_assignment.max_score
+
+    def _get_num_students(self):
+        with Gradebook("sqlite:////mnt/nbgrader/"+self._nbgrader_course+"/grader/gradebook.db") as gb:
+            return len(gb.students)
+
+    def _create_assignment(self, max_score):
+        return self._course.create_assignment({'name':self._form_nb_assign_name, 'published':'true', 'assignment_group_id':self._group, 'points_possible':max_score})
+        
+    def _delete_comments(self):
+        '''clears old comments from submission to prevent cluttering of the comment section in case of re-submission'''
+        update_status(self._form_nb_assign_name, self._course_id, completion=5, status='Removing Old Feedback')
+        counter = 0
+        auth_header = {'Authorization': 'Bearer ' + self._flask_session['api_key']}
+        with Gradebook("sqlite:////mnt/nbgrader/"+self._nbgrader_course+"/grader/gradebook.db") as gb:
+            for student in gb.students:
+                try:
+                    submission = self.assignment_to_upload.get_submission(self.canvas_students[student.id],include=['submission_comments'])
+                    
+                    for comment in submission.submission_comments:
+                        if comment['comment'] == 'Transferred from datahub.':
+                            response = requests.delete(
```

Comment:
Consider dependency injection with `requests` with `self._requests` by adding it like so

```python
class UploadGrades:

    def __init__(self, course_id, group, form_canvas_assign_id, form_nb_assign_name, late_penalty, lti=lti):
        self._requests = requests
```

That way, during testing, you can put it a fake request object to have it return whatever you want. See example code:

```python
from unittest.mock import MagicMock
import pytest

mock_requests = MagicMock()
mock_requests.delete = MagicMock(return_value=300)

def test_delete_comments_raises():
    with pytest.raises(SomeError):
        ug = UploadGrades()
        ug._requests = mock_requests
        ug._delete_comments()
```

See https://www.freecodecamp.org/news/a-quick-intro-to-dependency-injection-what-it-is-and-when-to-use-it-7578c84fa88f/ for more info